File: implementations/csharp/server/src/ShuaiTunnel.Server/PeerMesh/PeerDataFrameHeader.cs
using System.Buffers.Binary;

namespace ShuaiTunnel.Server.PeerMesh;

internal sealed record PeerDataFrameHeader(
    long SessionId,
    long Sequence)
{
    private const uint Magic = 0x53504d32; // SPM2
    private const int HeaderBytes = 20;
    private const int TagBytes = 16;
    private const int MaxFrameBytes = 65535;

    public static bool LooksLikeDataFrame(ReadOnlySpan<byte> frame) =>
        frame.Length >= 4 && BinaryPrimitives.ReadUInt32BigEndian(frame[..4]) == Magic;

    public static PeerDataFrameHeader? Parse(ReadOnlySpan<byte> frame)
    {
        if (frame.Length < HeaderBytes + TagBytes || frame.Length > MaxFrameBytes || !LooksLikeDataFrame(frame))
        {
            return null;
        }
        var sequence = (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(12, 8));
        return sequence > 0
            ? new PeerDataFrameHeader(
                (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(4, 8)),
                sequence)
            : null;
    }
}

File: implementations/java/common/src/main/java/com/theshuai/common/peermesh/PeerDataFrameHeader.java
package com.theshuai.common.peermesh;

import java.nio.ByteBuffer;

public record PeerDataFrameHeader(
        long sessionId,
        long sequence
) {
    private static final int MAGIC = 0x53504D32; // SPM2
    private static final int TAG_BYTES = 16;
    private static final int HEADER_BYTES = Integer.BYTES + Long.BYTES * 2;
    private static final int MIN_BYTES = HEADER_BYTES + TAG_BYTES;
    private static final int MAX_BYTES = 65_535;

    public static boolean looksLikeDataFrame(byte[] packet) {
        return packet != null
                && packet.length >= Integer.BYTES
                && ByteBuffer.wrap(packet, 0, Integer.BYTES).getInt() == MAGIC;
    }

    public static PeerDataFrameHeader parse(byte[] packet) {
        if (packet == null || packet.length < MIN_BYTES || packet.length > MAX_BYTES) {
            return null;
        }
        ByteBuffer buffer = ByteBuffer.wrap(packet);
        if (buffer.getInt() != MAGIC) {
            return null;
        }
        long sessionId = buffer.getLong();
        long sequence = buffer.getLong();
        return sessionId > 0 && sequence > 0
                ? new PeerDataFrameHeader(sessionId, sequence)
                : null;
    }
}
